spring-cloud-commons/src/main/java/org/springframework/cloud/commons/util/MismatchVerifierAutoConfiguration.java

```diff
@@ -0,0 +1,265 @@
+ */
+class SpringBootVersionVerifier implements MismatchVerifier {
+
+	private static final Log log = LogFactory.getLog(SpringBootVersionVerifier.class);
+
+	final Map<String, MismatchPredicate> ACCEPTED_VERSIONS = new HashMap<String, MismatchPredicate>() {
+		{
+			this.put("1.5", is1_5());
+			this.put("2.0", is2_0());
+			this.put("2.1", is2_1());
+		}
+	};
+
+	private final List<String> acceptedVersions;
+
+	SpringBootVersionVerifier(List<String> acceptedVersions) {
+		this.acceptedVersions = acceptedVersions;
+	}
+
+	@Override
+	public VerificationResult verify() {
+		boolean matches = springBootVersionMatches();
+		if (matches) {
+			return new VerificationResult();
+		}
+		return new VerificationResult(errorDescription());
+	}
+
+	MismatchPredicate is1_5() {
+		return new MismatchPredicate() {
+			@Override
+			public boolean accept() {
+				try {
+					// deprecated 1.5
+					Class.forName("org.springframework.boot.context.config.ResourceNotFoundException");
+					return true;
+				}
+				catch (ClassNotFoundException e) {
+					return false;
+				}
+			}
+		};
+	}
+
+	private boolean bootVersionFromManifest(String s) {
+		String version = getVersionFromManifest();
+		if (log.isDebugEnabled()) {
+			log.debug("Version found in Boot manifest [" + version + "]");
+		}
+		return StringUtils.hasText(version) && version.startsWith(s);
+	}
+
+	String getVersionFromManifest() {
+		return SpringBootVersion.getVersion();
+	}
+
+	MismatchPredicate is2_0() {
+		return new MismatchPredicate() {
+			@Override
+			public boolean accept() {
+				try {
+					// present in 2.0, 1.5 missing in 2.1
+					SpringApplicationBuilder.class.getMethod("web", boolean.class);
+					return !is1_5().accept();
+				}
+				catch (NoSuchMethodException e) {
+					return false;
+				}
+			}
+		};
+	}
+
+	MismatchPredicate is2_1() {
+		return new MismatchPredicate() {
+			@Override
+			public boolean accept() {
+				try {
+					// since 2.1
+					Class.forName("org.springframework.boot.task.TaskExecutorCustomizer");
+					return true;
+				}
+				catch (ClassNotFoundException e) {
+					return false;
+				}
+			}
+		};
+	}
+
+	private String errorDescription() {
+		return "You are using Spring Boot in version incompatible with compatible versions " + this.acceptedVersions;
+	}
+
+	private boolean springBootVersionMatches() {
+		for (String acceptedVersion : acceptedVersions) {
+			if (bootVersionFromManifest(acceptedVersion)) {
+				return true;
+			}
+			else {
+				// 2.0, 2.1
+				MismatchPredicate predicate = ACCEPTED_VERSIONS.get(acceptedVersion);
+				if (predicate != null && predicate.accept()) {
+					return true;
+				}
+			}
+		}
+		return false;
+	}
+}
+
+class VerificationResult {
+	final String description;
+
+	// if OK
+	VerificationResult() {
+		this.description = "";
+	}
+
+	// if not OK
+	VerificationResult(String errorDescription) {
+		this.description = errorDescription;
+	}
+}
+
+class DependencyMismatchReportMessage {
```

Comment:
I wonder if we could throw a specific exception and then use boot's failure analyzer mechanism rather than creating our own reporting mechanism. https://www.baeldung.com/spring-boot-failure-analyzer